tuxemon/states/items/__init__.py

```diff
@@ -393,36 +394,42 @@ def buy_item(quantity: int) -> None:
             if not quantity:
                 return
 
-            ## TODO: Allow money to transfer between buyer and seller.
-            if self.buyer:
-                self.seller.give_item(self.client, self.buyer, item, quantity)
-            else:
-                self.seller.alter_item_quantity(
-                    self.client,
-                    item.slug,
-                    -quantity,
-                )
+            self.buyer.buy_transaction(
+                self.client, self.seller, item.slug, quantity, price
+            )
+
             self.reload_items()
             if not self.seller.has_item(item.slug):
                 # We're pointing at a new item
                 self.on_menu_selection_change()
 
         item_dict = self.seller.inventory[item.slug]
-        max_quantity = (
-            None if item_dict.get("infinite") else item_dict["quantity"]
-        )
+
         price = (
-            1
+            0
             if not self.economy
             or not self.economy.lookup_item_price(item.slug)
             else self.economy.lookup_item_price(item.slug)
         )
+        money = self.buyer.game_variables.get("money")
+        qty_buyer_can_afford = (
```

Comment:
I think you should split this into two branches.

If price is 0 then max quantity is (some sensible maximum) if it's infinite, else just the amount the seller has.

If price is not zero then it is the minimum of what you can afford Vs how many they have.

Reply:
If they have 10000 items and price is zero you can buy 10000. But if it is infinite and price is zero you can only buy 999. Seems wrong.

Reply:
By 'splitting it into two branches', do you mean something like the following?
```
        MAX_QTY = 999

        if int(price) == 0:                                                                                                                                                                    
            if item_dict.get("infinite"):                                                                                                                                                      
                max_quantity = MAX_QTY                                                                                                                                                         
            else:                                                                                                                                                                              
                max_quantity =  min(MAX_QTY, item_dict["quantity"])                                                                                                                            
        else:                                                                                                                                                                                  
            qty_can_afford = int(money / price)                                                                                                                                                
            max_quantity = (                                                                                                                                                                   
                min(MAX_QTY, qty_can_afford, item_dict["quantity"])                                                                                                                       
            )                                                          ```

Reply:
Perfect